File: Semester3/Advanced Methods of Programming/Toy Language Interpreter C#/Assignement7/Assignement7/Collection/MyDictionary.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Assignement7.Collection
{
    public class MyDictionary<K, V> : MyIDictionary<K, V>
    {
        private Dictionary<K, V> map;

        public MyDictionary(Dictionary<K, V> map)
        {
            this.map = map;
        }

        public MyIDictionary<K, V> Clone()
        {
            return new MyDictionary<K, V>(new Dictionary<K, V>(this.map));
        }

        public V Get(K key)
        {
            if (this.map.ContainsKey(key))
                return this.map[key];
            else
                throw new Exception("Key not exist in table");
        }

        public List<K> Keys()
        {
            return new List<K>(map.Keys);
        }

        public List<V> Values()
        {
            return new List<V>(map.Values);
        }

        public void Put(K key, V value)
        {
            this.map[key] = value;
        }

        public void Remove(K key)
        {
            this.map.Remove(key);
        }

        public Dictionary<K, V> ToDictionary()
        {
            return this.map;
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            this.Keys().ForEach(e => sb.Append(e.ToString() + " -> " + this.Get(e).ToString() + "\n"));
            return sb.ToString();
        }
    }
}

File: Semester3/Advanced Methods of Programming/ExamenMAP/src/myCollections/MyDictionary.java
package myCollections;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class MyDictionary<K, V> implements MyIDictionary<K, V>{
    private HashMap<K, V> dictionary;

    public MyDictionary() {dictionary = new HashMap<>();}

    @Override
    public V get(K key) {
        return dictionary.get(key);
    }

    @Override
    public V put(K key, V value) {
        return dictionary.put(key, value);
    }

    @Override
    public int size() {
        return dictionary.size();
    }

    @Override
    public boolean containsKey(K key) {
        return dictionary.containsKey(key);
    }

    @Override
    public void remove(K key) {
        dictionary.remove(key);
    }

    @Override
    public Collection<V> values() {
        return dictionary.values();
    }

    @Override
    public boolean containsValue(V value) {
        return dictionary.containsValue(value);
    }

    @Override
    public Set<K> keySet() {
        return dictionary.keySet();
    }

    @Override
    public Set<Map.Entry<K, V>> entrySet() {
        return dictionary.entrySet();
    }

    @Override
    public MyIDictionary<K, V> clone_dict() {
        MyIDictionary<K, V> dict = new MyDictionary<>();
        for(K key : this.keySet())
            dict.put(key, dictionary.get(key));
        return dict;
    }

    @Override
    public Iterable<Map.Entry<K, V>> getAll() {
        return dictionary.entrySet();
    }

    @Override
    public String toString()
    {
        StringBuilder buffer = new StringBuilder();
        for(K key : this.keySet())
            buffer.append(key.toString()).append(" -> ").append(dictionary.get(key).toString()).append("\n");

        return buffer.toString();
    }
}
